File: src/backend/routers/messages.py
"""
Proactive Messaging Router (Issue #321).

Enables agents to send proactive messages to users across channels.
Authorization via allow_proactive flag on agent_sharing.
"""

import logging
from typing import Literal, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, EmailStr, Field

from database import db
from dependencies import get_current_user, AuthorizedAgent
from db_models import User
from services.proactive_message_service import (
    proactive_message_service,
    NotAuthorizedError,
    RecipientNotFoundError,
    RateLimitedError,
    ChannelDeliveryError,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/agents", tags=["messages"])


# =============================================================================
# Request/Response Models
# =============================================================================


class SendMessageRequest(BaseModel):
    """Request to send a proactive message to a user."""
    recipient_email: EmailStr = Field(
        ...,
        description="Verified email of the recipient. Must be in agent_sharing with allow_proactive=1."
    )
    text: str = Field(
        ...,
        min_length=1,
        max_length=4096,
        description="Message content (max 4096 characters)"
    )
    channel: Literal["auto", "telegram", "slack", "web"] = Field(
        default="auto",
        description="Target channel. 'auto' tries channels in order: telegram -> slack -> web"
    )
    reply_to_thread: bool = Field(
        default=False,
        description="Continue in last thread if one exists (channel-dependent)"
    )


class SendMessageResponse(BaseModel):
    """Response from sending a proactive message."""
    success: bool
    channel: str
    message_id: Optional[str] = None
    error: Optional[str] = None


class ProactiveShareUpdate(BaseModel):
    """Request to update allow_proactive flag for a share."""
    email: EmailStr
    allow_proactive: bool


class ProactiveSharesResponse(BaseModel):
    """List of emails with proactive messaging enabled."""
    agent_name: str
    emails: list[str]


# =============================================================================
# Endpoints
# =============================================================================


@router.post("/{agent_name}/messages", response_model=SendMessageResponse)
async def send_proactive_message(
    agent_name: str,
    request: SendMessageRequest,
    agent: AuthorizedAgent = Depends(),
):
    """
    Send a proactive message to a user from this agent.

    The recipient must:
    1. Be in agent_sharing for this agent with allow_proactive=1, OR
    2. Be the owner of the agent

    Rate limited to 10 messages per recipient per hour.
    """
    if agent.name != agent_name:
        raise HTTPException(status_code=403, detail="Agent name mismatch")

    try:
        result = await proactive_message_service.send_message(
            agent_name=agent_name,
            recipient_email=request.recipient_email,
            text=request.text,
            channel=request.channel,
            reply_to_thread=request.reply_to_thread,
        )

        return SendMessageResponse(
            success=result.success,
            channel=result.channel,
            message_id=result.message_id,
            error=result.error,
        )

    except NotAuthorizedError as e:
        raise HTTPException(status_code=403, detail=str(e))

    except RateLimitedError as e:
        raise HTTPException(status_code=429, detail=str(e))

    except RecipientNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))

    except ChannelDeliveryError as e:
        raise HTTPException(status_code=502, detail=str(e))

    except Exception as e:
        logger.exception(f"Proactive message failed: {e}")
        raise HTTPException(status_code=500, detail="Internal error sending message")


@router.put("/{agent_name}/shares/proactive", response_model=dict)
async def update_proactive_setting(
    agent_name: str,
    request: ProactiveShareUpdate,
    current_user: User = Depends(get_current_user),
):
    """
    Update the allow_proactive flag for a sharing record.

    Only the agent owner or admin can modify this setting.
    """
    success = db.set_allow_proactive(
        agent_name=agent_name,
        email=request.email,
        allow=request.allow_proactive,
        setter_username=current_user.username,
    )

    if not success:
        raise HTTPException(
            status_code=404,
            detail=f"Share not found or not authorized to modify"
        )

    return {
        "success": True,
        "agent_name": agent_name,
        "email": request.email,
        "allow_proactive": request.allow_proactive,
    }


@router.get("/{agent_name}/shares/proactive", response_model=ProactiveSharesResponse)
async def get_proactive_shares(
    agent_name: str,
    current_user: User = Depends(get_current_user),
):
    """
    List all emails that have opted in to receive proactive messages from this agent.

    Only the agent owner or admin can view this list.
    """
    if not db.can_user_share_agent(current_user.username, agent_name):
        raise HTTPException(status_code=403, detail="Not authorized to view shares")

    emails = db.get_proactive_enabled_shares(agent_name)

    return ProactiveSharesResponse(
        agent_name=agent_name,
        emails=emails,
    )

File: src/mcp-server/src/tools/messages.ts
/**
 * Proactive User Messaging Tools (Issue #321)
 *
 * MCP tool for agents to send proactive messages to specific users
 * by verified email across Telegram, Slack, and web channels.
 */

import { z } from "zod";
import { TrinityClient } from "../client.js";
import type { McpAuthContext } from "../types.js";

/**
 * Create message tools with the given client
 * @param client - Base Trinity client (provides base URL)
 * @param requireApiKey - Whether API key authentication is enabled
 */
export function createMessageTools(
  client: TrinityClient,
  requireApiKey: boolean
) {
  /**
   * Get Trinity client with appropriate authentication
   */
  const getClient = (authContext?: McpAuthContext): TrinityClient => {
    if (requireApiKey) {
      if (!authContext?.mcpApiKey) {
        throw new Error("MCP API key authentication required but no API key found in request context");
      }
      const userClient = new TrinityClient(client.getBaseUrl());
      userClient.setToken(authContext.mcpApiKey);
      return userClient;
    }
    return client;
  };

  /**
   * Get the agent name from auth context.
   * For agent-scoped keys, this is the bound agent.
   * For user-scoped keys, the agent must be specified in the call.
   */
  const getAgentName = (
    authContext: McpAuthContext | undefined,
    specifiedAgent?: string
  ): string => {
    // If an agent name is specified, use that (will be validated by backend)
    if (specifiedAgent) {
      return specifiedAgent;
    }
    // For agent-scoped keys, use the bound agent
    if (authContext?.scope === "agent" && authContext.agentName) {
      return authContext.agentName;
    }
    throw new Error(
      "Agent name is required. Either use an agent-scoped API key or specify the agent_name parameter."
    );
  };

  return {
    // ========================================================================
    // send_message - Send proactive message to a user by email
    // ========================================================================
    sendMessage: {
      name: "send_message",
      description:
        "Send a proactive message to a specific user by their verified email address. " +
        "The recipient must have opted in to receive proactive messages from this agent " +
        "(allow_proactive flag must be set in their sharing record). " +
        "Messages are delivered via Telegram, Slack, or web based on the channel parameter. " +
        "Rate limited to 10 messages per recipient per hour.",
      parameters: z.object({
        recipient_email: z.string().email()
          .describe(
            "The verified email address of the recipient. " +
            "Must be in agent_sharing with allow_proactive=1, or be the agent owner."
          ),
        text: z.string().min(1).max(4096)
          .describe("Message content to send (max 4096 characters)"),
        channel: z.enum(["auto", "telegram", "slack", "web"]).default("auto")
          .describe(
            "Target channel: 'auto' tries channels in order (telegram -> slack -> web), " +
            "or specify a specific channel. Currently Telegram and Slack are supported."
          ),
        reply_to_thread: z.boolean().default(false)
          .describe("Continue in the last thread with this user if one exists (channel-dependent)"),
        agent_name: z.string().optional()
          .describe(
            "Agent name to send as. Required for user-scoped API keys. " +
            "For agent-scoped keys, defaults to the calling agent."
          ),
      }),
      execute: async (
        params: {
          recipient_email: string;
          text: string;
          channel?: "auto" | "telegram" | "slack" | "web";
          reply_to_thread?: boolean;
          agent_name?: string;
        },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        // Validate message
        if (!params.text || params.text.trim().length === 0) {
          return JSON.stringify({
            success: false,
            error: "Message cannot be empty",
          }, null, 2);
        }

        if (params.text.length > 4096) {
          return JSON.stringify({
            success: false,
            error: "Message exceeds 4096 character limit",
          }, null, 2);
        }

        // Validate email format
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        if (!emailRegex.test(params.recipient_email)) {
          return JSON.stringify({
            success: false,
            error: "Invalid email address format",
          }, null, 2);
        }

        try {
          const agentName = getAgentName(authContext, params.agent_name);

          console.log(
            `[send_message] Sending to ${params.recipient_email} ` +
            `as ${agentName} via ${params.channel || "auto"} (${params.text.length} chars)`
          );

          const result = await apiClient.sendUserMessage(agentName, {
            recipient_email: params.recipient_email,
            text: params.text.trim(),
            channel: params.channel || "auto",
            reply_to_thread: params.reply_to_thread || false,
          });

          if (result.success) {
            return JSON.stringify({
              success: true,
              agent_name: agentName,
              recipient_email: params.recipient_email,
              channel: result.channel,
              message_id: result.message_id,
            }, null, 2);
          } else {
            return JSON.stringify({
              success: false,
              error: result.error || "Unknown error sending message",
            }, null, 2);
          }
        } catch (error) {
          const errorMessage = error instanceof Error ? error.message : String(error);
          console.error(`[send_message] Error: ${errorMessage}`);

          // Parse specific error types
          if (errorMessage.includes("403") || errorMessage.includes("Not authorized")) {
            return JSON.stringify({
              success: false,
              error: "Not authorized to message this recipient. They must opt in via allow_proactive flag.",
              not_authorized: true,
            }, null, 2);
          }

          if (errorMessage.includes("429") || errorMessage.includes("Rate limit")) {
            return JSON.stringify({
              success: false,
              error: "Rate limited. Max 10 messages per recipient per hour.",
              rate_limited: true,
            }, null, 2);
          }

          if (errorMessage.includes("404") || errorMessage.includes("not found")) {
            return JSON.stringify({
              success: false,
              error: "Recipient not found. They may not have Telegram or Slack configured.",
              recipient_not_found: true,
            }, null, 2);
          }

          return JSON.stringify({
            success: false,
            error: errorMessage,
          }, null, 2);
        }
      },
    },
  };
}
